File: CS208_AlgorithmDesignandAnalysis/Lab_Code/Lab7/B.cpp
#include <iostream>
using namespace std;

const int n = 2e5 + 1;
int arr[n];

struct node {
    long long sum, l_max, r_max, max;
} d[4 * n];

inline void bulid(int l, int r, int index);
inline void update(int index);
inline node merge(node p, node q);
inline node query(int l, int r, const int l_edge, const int r_edge, int index);

int main() {
    ios::sync_with_stdio(false);
    int num;
    cin >> num;
    for (size_t i = 1; i < num + 1; i++) {
        cin >> arr[i];
    }
    bulid(1, num + 1, 1);
    int m, l, r;
    cin >> m;
    for (size_t i = 0; i < m; i++) {
        cin >> l >> r;
        node ans = query(1, num + 1, l, r, 1);
        cout << ans.max << endl;
    }
}

inline void bulid(int l, int r, int index) {
    if (l == r) {
        d[index].l_max = d[index].r_max = d[index].max = d[index].sum = arr[l];
        return;
    }
    int mid = (r + l) / 2;
    bulid(l, mid, index * 2);
    bulid(mid + 1, r, index * 2 + 1);
    update(index);
}

inline void update(int index) {
    d[index].sum = d[index * 2].sum + d[index * 2 + 1].sum;
    d[index].l_max = max(d[index * 2].l_max, d[index * 2].sum + d[index * 2 + 1].l_max);
    d[index].r_max = max(d[index * 2 + 1].r_max, d[index * 2 + 1].sum + d[index * 2].r_max);
    d[index].max = max(max(d[index * 2].r_max + d[index * 2 + 1].l_max, d[index * 2].max), d[index * 2 + 1].max);
}

inline node merge(node p, node q) {  // p is left node, q is right node
    node tmp;
    tmp.sum = p.sum + q.sum;
    tmp.l_max = max(p.l_max, p.sum + q.l_max);
    tmp.r_max = max(q.r_max, q.sum + p.r_max);
    tmp.max = max(max(p.r_max + q.l_max, p.max), q.max);
    return tmp;
}

inline node query(int l, int r, const int l_edge, const int r_edge, int index) {
    if (l_edge <= l && r <= r_edge) {
        return d[index];
    }
    int mid = (l + r) / 2;
    if (l_edge <= mid && r_edge > mid) {
        node rr, ll;
        ll = query(l, mid, l_edge, r_edge, index * 2);
        rr = query(mid + 1, r, l_edge, r_edge, index * 2 + 1);
        return merge(ll, rr);
    } else if (l_edge > mid) {
        return query(mid + 1, r, l_edge, r_edge, index * 2 + 1);
    } else {
        return query(l, mid, l_edge, r_edge, index * 2);
    }
}

File: CS208_AlgorithmDesignandAnalysis/Lab_Code/Lab7/src/B.java
import java.io.*;
import java.util.Arrays;
import java.util.StringTokenizer;

public class B {
    private static class node {
        boolean l_edge;
        int l;
        int r;
        int index;
        node next;

        node(int ll, int rr, int ii, boolean edge) {
            l = ll;
            r = rr;
            index = ii;
            l_edge = edge;
        }
    }

    public static void main(String[] args) {
        QReader reader = new QReader();
        QWriter writer = new QWriter();
        int n = reader.nextInt();
        long[] arr = new long[n + 1];
        long[] sum = new long[n + 1];
        arr[0] = 0;
        sum[0] = 0;
        for (int i = 1; i < n + 1; i++) {
            arr[i] = reader.nextLong();
            sum[i] = sum[i - 1] + arr[i];
        }
        int m = reader.nextInt();
        node[] list = new node[n + 1];
        long[] ans = new long[m];
        Arrays.fill(ans, Long.MIN_VALUE);
        for (int i = 0; i < m; i++) {
            int p = reader.nextInt();
            int q = reader.nextInt();
            if (list[p] == null) {
                list[p] = new node(p, q, i, true);
            } else {
                node tmp = list[p];
                while (tmp.next != null) {
                    tmp = tmp.next;
                }
                tmp.next = new node(p, q, i, true);
            }
            if (list[q] == null) {
                list[q] = new node(p, q, i, false);
            } else {
                node tmp = list[q];
                while (tmp.next != null) {
                    tmp = tmp.next;
                }
                tmp.next = new node(p, q, i, false);
            }
        }
        query(1, n, arr, sum, list, ans);
        for (long l : ans) {
            writer.println(l);
        }
        writer.close();
    }

    public static long query(int l, int r, long[] arr, long[] sum, node[] list, long[] ans) {
        long __max;
        int mid = (l + r) / 2;
        if (l == r) {
            node tmp = list[l];
            while (tmp != null) {
                ans[tmp.index] = Math.max(ans[tmp.index], arr[l]);
                tmp = tmp.next;
            }
            __max = arr[l];
        } else {
            long l_max = query(l, mid, arr, sum, list, ans);
            long r_max = query(mid + 1, r, arr, sum, list, ans);
            long cross_max = merge(l, r, mid, sum);
            __max = Math.max(l_max, r_max);
            __max = Math.max(__max, cross_max);
            for (int i = l; i <= r; i++) {
                node tmp = list[i];
//                while (tmp != null) {
//                    if (l < tmp.l && tmp.l <= mid && mid < tmp.r && tmp.r < r){
//                        ans[tmp.index] = Math.max(ans[tmp.index], merge(tmp.l, tmp.r))
//                    } else if () {
//
//                    }else if (){
//
//                    }else if (tmp.l <= l && tmp.r >= r){
//                        ans[tmp.index] = Math.max(ans[tmp.index], __max);
//                    }
//                    tmp = tmp.next;
//                }
            }
        }
        return __max;
    }

    public static long merge(int l, int r, int mid, long[] sum) {//FIXME: TLE !!!
        long min = Long.MAX_VALUE;
        long max = Long.MIN_VALUE;
        for (int i = mid + 1; i <= r; i++) {
            if (max < sum[i]) {
                max = sum[i];
            }
        }
        for (int i = mid - 1; i >= l - 1; i--) {
            if (min > sum[i]) {
                min = sum[i];
            }
        }
        return max - min;
    }
}

class QReader {
    private BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
    private StringTokenizer tokenizer = new StringTokenizer("");

    private String innerNextLine() {
        try {
            return reader.readLine();
        } catch (IOException e) {
            return null;
        }
    }

    public boolean hasNext() {
        while (!tokenizer.hasMoreTokens()) {
            String nextLine = innerNextLine();
            if (nextLine == null) {
                return false;
            }
            tokenizer = new StringTokenizer(nextLine);
        }
        return true;
    }

    public String nextLine() {
        tokenizer = new StringTokenizer("");
        return innerNextLine();
    }

    public String next() {
        hasNext();
        return tokenizer.nextToken();
    }

    public int nextInt() {
        return Integer.parseInt(next());
    }

    public long nextLong() {
        return Long.parseLong(next());
    }
}

class QWriter implements Closeable {
    private BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));

    public void print(Object object) {
        try {
            writer.write(object.toString());
        } catch (IOException e) {
            return;
        }
    }

    public void println(Object object) {
        try {
            writer.write(object.toString());
            writer.write("\n");
        } catch (IOException e) {
            return;
        }
    }

    @Override
    public void close() {
        try {
            writer.close();
        } catch (IOException e) {
            return;
        }
    }
}